File: 907. Sum of Subarray Minimums/SumSubarrayMins.py
'''
https://leetcode.com/problems/sum-of-subarray-minimums/
'''


from typing import List


class Solution:
    # O(n) time,
    # O(n) space
    def sumSubarrayMins(self, arr: list[int]) -> int:
        n = len(arr)
        ans = 0
        
        inc_stack = []
        
        def pushToStack(index:int, stack:list) -> int:
            tot = 0
            num = arr[index]
            while stack and num < arr[stack[-1]]:
                curr_i = stack.pop()
                prev_i = stack[-1] if stack else -1
                tot += (curr_i - prev_i) * (i - curr_i) * (arr[curr_i])
                
            stack.append(i)
            
            return tot
        
        for i in range(n):
            ans += pushToStack(i, inc_stack)
         
        stack = inc_stack
        i = n
        while stack:
            curr_i = stack.pop()
            prev_i = stack[-1] if stack else -1
            ans += (curr_i - prev_i) * (i - curr_i) * (arr[curr_i])
                
        return ans % (10**9 + 7)

    
    # O(n) time,
    # O(n) space,
    # Approach: monotonic stack, arra
    def sumSubarrayMins2(self, arr: List[int]) -> int:
        
        def findSubarrayMinSum(nums: List[int]) -> int:
            tot = 0
            stack = [] # monotonic(non decreasing) stack
            
            for index, num in enumerate(nums):
                while stack and num <= nums[stack[-1]]:
                    popped_index = stack.pop()
                    left_offset = stack[-1] if stack else -1
                    right_offset = index
                    num_subarrays = (popped_index-left_offset) * (right_offset-popped_index)
                    tot += nums[popped_index] * num_subarrays
                stack.append(index)
            
            right_offset = len(nums)
            while stack:
                popped_index = stack.pop()
                left_offset = stack[-1] if stack else -1
                num_subarrays = (popped_index-left_offset) * (right_offset-popped_index)
                tot += nums[popped_index] * num_subarrays
                
            return tot
        
        return findSubarrayMinSum(arr) % (10**9 + 7)


sol = Solution()
print(sol.sumSubarrayMins([3,1,2,4]))

File: 907. Sum of Subarray Minimums/sum_subarray_mins.rs

impl Solution {
    // O(n) time,
    // O(n) space,
    // Approach: monotonic stack, 
    pub fn sum_subarray_mins(arr: Vec<i32>) -> i32 {
       
        let MOD = i64::pow(10, 9) + 7;
        let mut mins_index: Vec<usize> = Vec::new();
        let mut sum = 0 as i64;
        
        for index in 0..arr.len() {
            let num = arr[index];
            
            while mins_index.len() > 0 && num < arr[*mins_index.last().unwrap()] {
                
                let curr_index = mins_index.pop().unwrap();
                let curr_min = arr[curr_index];
                let left = if mins_index.len() > 0 { *mins_index.last().unwrap() as i32} else { -1 };
                let right = index;
                
                let right_sum = ((right-curr_index) as i32);
                let left_sum = (curr_index as i32 - left);
                sum += ((left_sum * right_sum) as i64 * curr_min as i64) % MOD;
                // sum %= MOD;
            }
            
            mins_index.push(index);
        }
        
        // println!("{:?}", mins_index);
        let right = arr.len();
        while mins_index.len() > 0 {
                
                let curr_index = mins_index.pop().unwrap();
                let curr_min = arr[curr_index];
                let left = if mins_index.len() > 0 { *mins_index.last().unwrap() as i32} else { -1 };
                
                let right_sum = ((right-curr_index) as i32);
                let left_sum = (curr_index as i32 -left);
            
                sum += ((left_sum * right_sum) as i64 * curr_min as i64) % MOD;
                // sum %= MOD;
        }
        
        return (sum % MOD) as i32;
    }
}